DDF-3763 Added fabric-karaf-checks to boot features

oconnormi opened the pull request:
#### What does this PR do?

Adds the health and readiness endpoints from fabric8 to the ddf boot features

 * supports more intelligent load balancing via health monitoring
 * supports readiness checks that help determine when it is ok to interact with the system
 * extensible to support additional health and readiness checks if needed

Not sure if the way I'm adding it is the best way to do this, open to suggestions.

#### Who is reviewing it? 

@tbatie @Lambeaux 

#### Choose 2 committers to review/merge the PR. 

@lessarderic @pklinef 

#### How should this be tested? (List steps with links to updated documentation)

* Start the system, 
ping https://localhost:8993/readiness-check and https://localhost:8993/health-check until they return `200`

#### Any background context you want to provide?

https://fabric8.io/guide/karaf.html#fabric8-karaf-health-checks

#### What are the relevant tickets?
[DDF-3763](https://codice.atlassian.net/browse/DDF-3763)
#### Screenshots (if appropriate)
#### Checklist:
- [ ] Documentation Updated
- [x] Add feature to security manager policy

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

pklinef commented:
The custom health checks will be very useful for our itests.
https://fabric8.io/guide/karaf.html#add-custom-heath-checks

oconnormi commented:
I updated this to be included in the kernel features so it should be started as one of the first few boot features. I wanted to check with @tbatie to see if this was the best way or not.

Also added permissions to the security manager policy. 

Still need to add something to the documentation. I wasn't sure if i should make a mention of it in the developer docs as well as the admin docs since it allows for extensibility, any thoughts?

tbatie commented on features/kernel/src/main/resources/features.xml:
```diff
@@ -81,4 +82,10 @@
         <bundle>wrap:mvn:net.jodah/failsafe/${jodah-failsafe.version}</bundle>
     </feature>
+
+    <feature name="healthchecks" description="Health and Readiness Checks" version="${fabric8.version}">
```
❓ Is there a reason we didn't use their features file? https://github.com/fabric8io/fabric8/blob/master/components/fabric8-karaf/fabric8-karaf-features/src/main/resources/feature.xml

oconnormi replied:
I think only because I didn't know if we wanted the other features that were in it

tbatie replied:
It doesn't look like they're adding any additional dependencies in those features that we don't already have. I'd be a fan of having our karaf features file have a repository tag that points to their feature file. One less feature for us to maintain 😄

oconnormi replied:
Yeah we can definitely do that, just wasn't sure if we wanted to pull in this stuff as well: https://github.com/fabric8io/kubernetes-client/blob/master/platforms/karaf/features/src/main/resources/feature.xml since it is referenced in that `fabric8-karaf` feature repo there shouldn't be a problem with pulling all that in AFAICT, just wasn't sure if we needed it at the moment

oconnormi commented:
Build Now

LinkMJB commented:
build now (testing a new pre-step change)

oconnormi commented on features/kernel/src/main/feature/feature.xml:
```diff
@@ -81,4 +82,10 @@
         <bundle>wrap:mvn:net.jodah/failsafe/${jodah-failsafe.version}</bundle>
     </feature>
+
+    <feature name="healthchecks" description="Health and Readiness Checks" version="${fabric8.version}">
```
Just changed this to pull in their feature repo

oconnormi commented:
build now

oconnormi commented:
Alright, think I finally fixed the itest issues

Build Now

paouelle commented on distribution/ddf-common/src/main/resources/security/default.policy:
```diff
@@ -115,7 +115,7 @@ grant codeBase "file:/registry-federation-admin-impl" {
 }
 
-grant codeBase "file:/spatial-csw-source/spatial-csw-endpoint/org.apache.camel.camel-core/com.google.guava/ddf-pubsub/registry-source-configuration-handler/registry-api-impl/org.apache.camel.camel-blueprint" {
+grant codeBase "file:/spatial-csw-source/spatial-csw-endpoint/org.apache.camel.camel-core/com.google.guava/ddf-pubsub/registry-source-configuration-handler/registry-api-impl/org.apache.camel.camel-blueprint/fabric8-karaf-checks" {
```
❓ Do we really want to give read access to the keystores to fabric8?

oconnormi replied:
I had to in order to get the itests to actually pass, they were running into access violations and the bundle wasn't starting because of that

paouelle replied:
That means we are exposing this in production to address an itests problem. 
@stustison What is your take on that?

oconnormi replied:
I'm not sure why I don't run into this problem when running the actual system and granting only the `createClassLoader` permission to the health check bundle. Everything seems to be working for me outside of itests in that case...

stustison replied:
I definitely don't like it. It might be time to explore having an itest permission file that we can drop in just for the tests when stuff like this happens.

oconnormi replied:
I can go with that approach if that will be preferred @stustison. Would we want it to be inherited by downstream projects automatically?

oconnormi replied:
Yeah so did some more testing and these are definitely required permissions for this feature to work

oconnormi commented:
Trying again, i think something happened when those tests tried to resolve artifacts from central

Build Now

oconnormi commented:
Build Now

oconnormi commented:
not sure what happened there, trying again. 

Build Now

oconnormi commented:
Build Now

lessarderic commented:
Marked CI build as successful since the only remaining problem was a known OWASP issue not related to this PR.
